src/components/primitives/mux-video.tsx

```diff
@@ -1,47 +1,86 @@
-import Hls, { Events } from 'hls.js'
+import type THls from 'hls.js'
 import * as React from 'react'
 import mergeRefs from 'react-merge-refs'
 
+import { createObserver } from '~/hooks/use-intersection-observer'
+
 export type MuxVideoProps = {
   muxSrc: string
+  lazy?: boolean
+  lazyObserverOptions?: IntersectionObserverInit
 } & Omit<JSX.IntrinsicElements['video'], 'src'>
 
+const supportsHls = (videoElm: HTMLVideoElement) =>
+  videoElm.canPlayType('application/vnd.apple.mpegurl')
+
 export const MuxVideo = React.forwardRef<HTMLVideoElement, MuxVideoProps>(
-  ({ muxSrc, className, ...rest }, ref) => {
+  ({ muxSrc, lazy = true, lazyObserverOptions, className, ...rest }, ref) => {
+    const hls = React.useRef<THls | null>(null)
     const videoRef = React.useRef<HTMLVideoElement>(null)
 
-    React.useEffect(() => {
-      let hls: Hls
+    const loadVideo = React.useCallback(
+      (videoElm: HTMLVideoElement) => {
+        if (hls.current) {
+          hls.current?.loadSource(muxSrc)
+        } else if (videoElm && supportsHls(videoElm)) {
+          videoElm.src = muxSrc
+        } else {
+          console.error('Unable to reproduce video')
+        }
+      },
+      [muxSrc]
+    )
 
-      if (videoRef.current) {
-        const video = videoRef.current
+    const importHls = React.useCallback(
+      async (videoElm: HTMLVideoElement) => {
+        const Hls = (await import('hls.js')).default
 
-        if (video.canPlayType('application/vnd.apple.mpegurl')) {
-          // Some browers (safari and ie edge) support HLS natively
-          video.src = muxSrc
-          video.defaultMuted = true
-        } else if (Hls.isSupported()) {
+        if (Hls.isSupported() && videoElm) {
           // This will run in all other modern browsers
-          hls = new Hls()
-          hls.attachMedia(video)
-          hls.on(Events.MEDIA_ATTACHED, () => {
-            hls?.loadSource(muxSrc)
-            hls?.on(Events.MANIFEST_PARSED, (_event, data) => {
+          hls.current = new Hls()
+          hls.current.attachMedia(videoElm)
+          hls.current.on(Hls.Events.MEDIA_ATTACHED, () => {
+            !lazy && loadVideo(videoElm)
+
+            hls.current?.on(Hls.Events.MANIFEST_PARSED, (_event, data) => {
               // Maximum quality available
-              hls.nextLevel ??= data.levels.length - 1
+              hls.current!.nextLevel ??= data.levels.length - 1
             })
           })
         } else {
           console.error("This is a legacy browser that doesn't support MSE")
         }
+      },
+      [lazy, loadVideo]
+    )
+
+    React.useEffect(() => {
+      if (!videoRef.current) return
+
+      const video = videoRef.current
+      let intersectionObserver: IntersectionObserver
+
+      if (video.canPlayType('application/vnd.apple.mpegurl')) {
```

Comment:
we can use the `supportsHlsNatively` util here

Reply:
I fixed this locally but didn´t push, lol.